File: customview/src/main/java/com/example/customview/activity/HoverItemActivity3.java
package com.example.customview.activity;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.TextView;

import com.example.customview.R;
import com.example.customview.adapter.HoverItemAdapter3;
import com.example.customview.bean.HoverItemModel;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

public class HoverItemActivity3 extends AppCompatActivity {

    public static final int MODEL_COUNT = 40;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_hover_item_3);

        RecyclerView mRecyclerView = (RecyclerView) findViewById(R.id.recyclerview);
        final TextView tvStickyHeaderView = (TextView) findViewById(R.id.header_view);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(this));
        mRecyclerView.setAdapter(new HoverItemAdapter3(this, getData()));
        mRecyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(@NotNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                View stickyInfoView = recyclerView.findChildViewUnder(tvStickyHeaderView.getMeasuredWidth() >> 1, 1);
                if (stickyInfoView != null) {
                    tvStickyHeaderView.setText(stickyInfoView.getContentDescription().toString());
                }

                View transInfoView = recyclerView.findChildViewUnder(tvStickyHeaderView.getMeasuredWidth() >> 1, tvStickyHeaderView.getMeasuredHeight() + 1);
                if (transInfoView != null) {
                    int transViewStatus = (int) transInfoView.getTag();
                    int dealtY = transInfoView.getTop() - tvStickyHeaderView.getMeasuredHeight();

                    if (transViewStatus == HoverItemAdapter3.HAS_STICKY_VIEW && transInfoView.getTop() > 0) {
                        tvStickyHeaderView.setTranslationY(dealtY);
                    } else {
                        tvStickyHeaderView.setTranslationY(0);
                    }
                }
            }
        });
    }

    public List<HoverItemModel> getData() {
        List<HoverItemModel> hoverItemModels = new ArrayList<>();
        for (int index = 0; index < MODEL_COUNT; index++) {
            if (index < 6) {
                hoverItemModels.add(new HoverItemModel("吸顶文本1", "name:" + index));
            } else if (index < 12) {
                hoverItemModels.add(new HoverItemModel("吸顶文本2", "name:" + index));
            } else if (index < 18) {
                hoverItemModels.add(new HoverItemModel("吸顶文本3", "name:" + index));
            } else {
                hoverItemModels.add(new HoverItemModel("吸顶文本4", "name:" + index));
            }
        }
        return hoverItemModels;
    }
}

File: customview/src/main/java/com/example/customview/activity/HoverItemActivity3.kt
package com.example.customview.activity

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.customview.adapter.HoverItemAdapter3
import com.example.customview.bean.HoverItemModel
import com.example.customview.databinding.ActivityHoverItem3Binding

/**
 * 封装 Hover Item Activity 3 相关逻辑的类。
 */
open class HoverItemActivity3 : AppCompatActivity() {
    private lateinit var binding: ActivityHoverItem3Binding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityHoverItem3Binding.inflate(layoutInflater)
        setContentView(binding.root)

        val mRecyclerView = binding.recyclerview
        val tvStickyHeaderView = binding.headerView
        mRecyclerView.layoutManager = LinearLayoutManager(this)
        mRecyclerView.adapter = HoverItemAdapter3(this, getData())
        mRecyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                recyclerView.findChildViewUnder((tvStickyHeaderView.measuredWidth shr 1).toFloat(), 1f)
                    ?.let { tvStickyHeaderView.text = it.contentDescription.toString() }

                val transInfoView = recyclerView.findChildViewUnder(
                    (tvStickyHeaderView.measuredWidth shr 1).toFloat(),
                    (tvStickyHeaderView.measuredHeight + 1).toFloat()
                )
                if (transInfoView != null) {
                    val transViewStatus = transInfoView.tag as Int
                    val dealtY = transInfoView.top - tvStickyHeaderView.measuredHeight
                    tvStickyHeaderView.translationY =
                        if (transViewStatus == HoverItemAdapter3.HAS_STICKY_VIEW && transInfoView.top > 0) {
                            dealtY.toFloat()
                        } else {
                            0f
                        }
                }
            }
        })
    }

    fun getData(): List<HoverItemModel> =
        (0 until MODEL_COUNT).map { index ->
            val sticky: String = when {
                index < 6 -> "吸顶文本1"
                index < 12 -> "吸顶文本2"
                index < 18 -> "吸顶文本3"
                else -> "吸顶文本4"
            }
            HoverItemModel(sticky, "name:$index")
        }


    companion object {
        const val MODEL_COUNT = 40
    }
}
